cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -453,29 +323,172 @@ angular.module(PKG.name + '.commons')
       };
     }
 
-    function formatConnections() {
-      var connections = [];
-      angular.forEach(vm.instance.getConnections(), function (conn) {
-        connections.push({
-          from: conn.sourceId,
-          to: conn.targetId
-        });
+    function addConnection(newConnObj) {
+      $scope.connections.push({
+        from: newConnObj.sourceId,
+        to: newConnObj.targetId
+      });
+      DAGPlusPlusNodesActionsFactory.setConnections($scope.connections);
+    }
+
+    function removeConnection(detachedConnObj, updateStore = true) {
+      var detachedConnection = _.find($scope.connections, function (conn) {
+        return conn.from === detachedConnObj.sourceId && conn.to === detachedConnObj.targetId;
+      });
+      if (detachedConnection) {
+        $scope.connections.splice($scope.connections.indexOf(detachedConnection), 1);
+      }
+      if (updateStore) {
+        DAGPlusPlusNodesActionsFactory.setConnections($scope.connections);
+      }
+    }
+
+    function moveConnection(moveInfo) {
+      let oldConnection = {
+        sourceId: moveInfo.originalSourceId,
+        targetId: moveInfo.originalTargetId
+      };
+      // don't need to call addConnection for the new connection, since that will be done
+      // automatically as part of the 'connection' event
+      removeConnection(oldConnection, false);
+    }
+
+    vm.removeSelectedConnections = function() {
+      if (selectedConnections.length === 0 || vm.isDisabled) { return; }
+
+      angular.forEach(selectedConnections, function (selectedConnectionObj) {
+        vm.instance.unbind('connectionDetached');
```

Comment:
Yeah, because `connectionDetached` will call the `removeConnection` function, which will update the store by default. When we're removing multiple connections programatically (e.g. in a loop), we don't want to update the store every time we remove one, as that will mess up the undo states. So we unbind `connectionDetached`, calls `removeConnection` manually with the `updateStore` param set to `false`, then binds `connectionDetached` again, so that it will work correctly when the user manually detaches a connection.